### PR TITLE
Complete Functor exercises

### PR DESCRIPTION
**Pull request by austinbutler**

### CODE REVIEW

File: src/Course/Functor.hs [hunk 1]
```diff
@@ -10,119 +11,130 @@ import Course.List
...
-    (<$>) = error "todo: Course.Functor (<$>)#instance ExactlyOne"
-
--- | Maps a function on the List functor.
---
--- >>> (+1) <$> Nil
--- []
---
--- >>> (+1) <$> (1 :. 2 :. 3 :. Nil)
--- [2,3,4]
+    (<$>) f (ExactlyOne x) = ExactlyOne $ f x
```
Comment (friedbrice) [inline src/Course/Functor.hs]:
```suggestion
    f <$> ExactlyOne x = ExactlyOne $ f x
```

This lets you avoid some brackets :-)

File: src/Course/Functor.hs [hunk 2]
```diff
@@ -10,119 +11,130 @@ import Course.List
...
-    (<$>) = error "todo: Course.Functor (<$>)#instance List"
-
--- | Maps a function on the Optional functor.
---
--- >>> (+1) <$> Empty
--- Empty
---
--- >>> (+1) <$> Full 2
--- Full 3
+    (<$>) f = foldRight (\y acc -> f y :. acc) Nil
```
Comment (friedbrice) [inline src/Course/Functor.hs:54]: Nice. I like how you reduced away the second argument :-)

File: fp-course.cabal [hunk 3]
```diff
@@ -5,7 +5,7 @@ version:              0.1.4
 build-type:           Simple
 extra-source-files:   etc/CONTRIBUTORS
 
-license:              BSD3
+license:              BSD-3-Clause
```
Comment (austinbutler) [inline fp-course.cabal:8]: Cabal complains about `BSD3`, see: https://github.com/haskell/cabal/issues/5109

May want to pick this up in your fork @friedbrice.

Comment (friedbrice) [reply fp-course.cabal:8]: I saw that! Thanks for bringing it up.

File: src/Course/Functor.hs [hunk 4]
```diff
@@ -10,119 +11,130 @@ import Course.List
...
+
+>>> (*2) :. (+1) :. const 99 :. Nil ?? 8
+[16,9,99]
+
+>>> Empty ?? 2
+Empty
+-}
 (??) :: Functor k => k (a -> b) -> a -> k b
-(??) ff a = error "todo: Course.Functor#(??)"
+(??) f a = (<$>) (\x -> x a) f
```
Comment (friedbrice) [inline src/Course/Functor.hs]:
```suggestion
(??) f a = (<$>) ($ a) f
```

> Avoid lambdas at all cost!

^ Haskell slogan ;-)

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
+{
```
Comment (austinbutler) [inline flake.nix:1]: Aside from what I'm calling out below this is a pretty simple, vanilla Nix Flake, so if you want to throw it into the main repo @friedbrice this should take care of Nix users. They can just `nix develop` or set up direnv.

File: flake.nix [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+          packages = p: [p.${packageName}];
+          buildInputs = with pkgs; [
+            alejandra
+            haskellPackages.cabal-install
+            haskellPackages.ghcid
+            haskellPackages.haskell-language-server
+            pre-commit
+          ];
+          shellHook = ''
+            pre-commit install
```
Comment (austinbutler) [inline flake.nix:46]: Don't need this, I just like being able to have consistent formatting and such.

File: flake.nix [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        packages.default = pkgs.haskellPackages.${packageName};
+
+        devShells.default = pkgs.haskellPackages.shellFor {
+          packages = p: [p.${packageName}];
+          buildInputs = with pkgs; [
+            alejandra
+            haskellPackages.cabal-install
+            haskellPackages.ghcid
+            haskellPackages.haskell-language-server
+            pre-commit
```
Comment (austinbutler) [inline flake.nix:43]: Don't need this, I just like being able to have consistent formatting and such.

File: flake.nix [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+          inherit config system;
+          overlays = [overlay];
+        };
+      in {
+        packages.default = pkgs.haskellPackages.${packageName};
+
+        devShells.default = pkgs.haskellPackages.shellFor {
+          packages = p: [p.${packageName}];
+          buildInputs = with pkgs; [
+            alejandra
```
Comment (austinbutler) [inline flake.nix:39]: Don't need this, it's just for formatting Nix files.

File: src/Course/Functor.hs [hunk 9]
```diff
@@ -10,119 +11,130 @@ import Course.List
...
+
+>>> void Empty
+Empty
+
+>>> void (+10) 5
+()
+-}
 void :: Functor k => k a -> k ()
-void = error "todo: Course.Functor#void"
+void = (<$>) (const ())
```
Comment (friedbrice) [inline src/Course/Functor.hs:128]: Good work on that. You could also implement this in terms of `<$`

**Review by friedbrice — APPROVED** [review]
Great work :-)

File: src/Course/Functor.hs (hunk 4 above)
Comment (austinbutler) [reply src/Course/Functor.hs]: Wow I never even though of using `$` that way...

**Pull request merged**